core/src/main/scala/kafka/log/Log.scala

```diff
@@ -444,6 +457,23 @@ class Log(@volatile var dir: File,
     }
   }
 
+  /*
+   * Increment the log start offset if the provided offset is larger.
+   * And delete old segments whose largest offset < log start offset.
+   */
+  def maybeIncrementLogStartOffsetAndDeleteRecords(offset: Long) {
```

Comment:
Good point. I will only delete non-compact topics in `maybeIncrementLogStartOffsetAndDeleteRecords` and let `LogCleaner` take care of compact topics.